### PR TITLE
Re-fetch order after payment

### PR DESCRIPTION
**Pull request by malinajirka**
Parent issue #3726 

The app automatically re-fetches order when the payment flow completes and refreshes state of the order detail screen. This is needed so the order status and "paid state" gets refreshed.

The logic is following
1. When the payment completes -> PaymentSuccessful screen is shown
2. The app automatically starts fetching the order
3a. If the user leaves PaymentSuccessful screen before the re-fetch completes a progress dialog is shown
3b. If the user leaves PaymentSuccessful screen after the re-fetch completes the progress dialog is NOT shown

Note: I don't like the unit tests at all. Especially the usage of @VisibleForTesting annotation. However, I'm not sure if there is any other way how to test this, since we need to simulate that "re-fetch" takes some time. Let me know if you have any better suggestions ;).

To Test:
1. Open a detail of an order in US dollars 
2. Go through the payment flow
3. Use charles to suspend re-fetch order request (or add a delay [here](https://github.com/woocommerce/woocommerce-android/blob/3beacefb0494cccbd667b6410cd1bc6c921a5b89/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt#L150))
4. Press back to dismiss PaymentSuccessful screen
5. Notice a progress dialog is shown
6. Resume re-fetch order request
7. Notice the progress gets dismissed and order detail gets updated = it is "Paid"
-------------------------
Run the same test again, but dismiss PaymentSuccessful screen after the request completes -> the progress should not be shown at all.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt [hunk 1]
```diff
@@ -134,15 +140,7 @@ class CardReaderPaymentViewModel @Inject constructor(
             CollectingPayment -> viewState.postValue(CollectPaymentState(amountLabel))
             ProcessingPayment -> viewState.postValue(ProcessingPaymentState(amountLabel))
             CapturingPayment -> viewState.postValue(CapturingPaymentState(amountLabel))
-            // TODO cardreader store receipt data into a persistent storage
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:137]: This TODO is gone now. Is this on purpose?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt [hunk 2]
```diff
@@ -576,6 +576,19 @@ class OrderDetailViewModel @Inject constructor(
         viewState = viewState.copy(refreshedProductId = event.remoteProductId)
     }
 
+    fun onCardReaderPaymentCompleted() {
+        reloadOrderDetails()
+    }
+
+    private fun reloadOrderDetails() {
+        launch {
+            orderDetailRepository.getOrder(navArgs.orderId)?.let {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:587]: If the order is missing in the DB then we just show the old order, correct? Seems to be like an inconsistent state situation, so maybe at least lets log it?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt [hunk 3]
```diff
@@ -576,6 +576,19 @@ class OrderDetailViewModel @Inject constructor(
         viewState = viewState.copy(refreshedProductId = event.remoteProductId)
     }
 
+    fun onCardReaderPaymentCompleted() {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:581]: Maybe at least some basic test for this code would be helpful?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt [hunk 4]
```diff
@@ -153,6 +151,26 @@ class CardReaderPaymentViewModel @Inject constructor(
...
+            { onPrintReceiptClicked(paymentStatus.receiptPaymentInfo) },
+            { onSendReceiptClicked(paymentStatus.receiptPaymentInfo) }
+        ))
+        reFetchOrder()
+    }
+
+    @VisibleForTesting
+    fun reFetchOrder() {
+        fetchOrderJob = launch {
+            orderRepository.fetchOrder(arguments.orderIdentifier)
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: If the order fetching fails we don't notify the user anyhow?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialog.kt [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package com.woocommerce.android.ui.orders.cardreader
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialog.kt:1]: I see that in the design there is a button "back to order".  WIll it be another PR or you forgot to add this?

image

**Review by kidinov — COMMENTED** [review]
Looks and works well! Just a couple of small things.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 4 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: It's pretty important btw, because in this case a user will end up on "processing" state of the order and "collect button" still will be active

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:137]: It wasn't - good catch! ;)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:587]: Good call ;)!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialog.kt (hunk 5 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentDialog.kt:1]: It wasn't in the designs at the time of implementation of this screen. We can fix this during the "polish ui" phase.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 4 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: I agree that at least showing an error message sounds like a good idea. I'm also thinking that we could hide the collect button when a receipt url is stored in the db, wdyt? Or would that logic be just confusing?

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: >  I'm also thinking that we could hide the collect button when a receipt url is stored in the db, wdyt? Or would that logic be just confusing?

Hm, now when you said that I got a thought. So if I got it right you update locally stored order with receipt url anyway. So wouldn't it make sense to actually update locally stored order and make it in "paid" state after we actually got confirmation that it's paid? In the background, we can still run "sync" with the backend but then we don't need to show loading dialog at all. Don't know the context so maybe it doesn't make sense at all though =).

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt (hunk 3 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:581]: Good call, I missed that - fixed in e72a997e1cc923534e07ab37c2c67681f316883c

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:137]: Fixed in 1ca8e819c3c2e807e9ed68f5c0754917b62fa029

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:587]: Fixed in 9f75388507131888045732c4af2257bb2511835e

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 4 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: Good thinking, unfortunately, the logic of changing states of an order is not straightforward - sometimes the order doesn't go into the completed state even when it's paid. 
Moreover, changing the state of an order locally can be quite tricky as the app won't know if it should push the updated state to the remote or if it's just a temporary local change. I plan to add the receipt url into a separate (localonly) table or perhaps even consider storing it in shared preferences. I think we'll want to store receipt url in the backend anyway so this is very probably just a temporary solution.

Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: Got it. Makes sense

>  I'm also thinking that we could hide the collect button when a receipt url is stored in the db, wdyt? Or would that logic be just confusing?

Back to your question. Yeah logic a bit confusing, but I think that "collect button" visible when we know that it was paid is bad as well. On debug/simulated build I even managed to pay twice for the same order, I hope it can not happen with real money though

Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: Fixed in ec3b7d972650250c9e8f25a0361a35c9ecc52f67

**Comment by malinajirka** [comment]
Thanks for the great review! I've implemented all your suggestions ;)! It's ready for another round.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt (hunk 4 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/cardreader/CardReaderPaymentViewModel.kt:168]: > On debug/simulated build I even managed to pay twice for the same order, I hope it can not happen with real money though

Actually, it can get quite tricky - if we create a new PaymentIntent (which we will), the user might actually get charged twice. I'll update the payment flow so it fetches the current order state at the beginning of the payment flow and fails the whole flow if the state can't be fetched, wdyt?
I added an item to https://github.com/woocommerce/woocommerce-android/issues/3726

**Review by kidinov — APPROVED** [review]
Looks good!
